vms/platformvm/block/executor/verifier.go

```diff
@@ -565,5 +569,73 @@ func (v *verifier) processStandardTxs(txs []*txs.Tx, feeCalculator fee.Calculato
 		}
 	}
 
+	// After processing all the transactions, deactivate any SoVs that might not
+	// have sufficient fee to pay for the next second.
+	//
+	// This ensures that SoVs are not undercharged for the next second.
+	err := deactivateLowBalanceSoVs(
+		v.txExecutorBackend.Config.ValidatorFeeConfig,
+		diff,
+	)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to deactivate low balance SoVs: %w", err)
+	}
+
 	return inputs, atomicRequests, onAcceptFunc, nil
 }
+
+// deactivateLowBalanceSoVs deactivates any SoVs that might not have sufficient
+// fees to pay for the next second.
+func deactivateLowBalanceSoVs(
+	config validatorfee.Config,
+	diff state.Diff,
+) error {
+	var (
+		accruedFees       = diff.GetAccruedFees()
+		validatorFeeState = validatorfee.State{
+			Current: gas.Gas(diff.NumActiveSubnetOnlyValidators()),
+			Excess:  diff.GetSoVExcess(),
+		}
+		potentialCost = validatorFeeState.CostOf(
+			config,
+			1, // 1 second
+		)
+	)
+	potentialAccruedFees, err := math.Add(accruedFees, potentialCost)
+	if err != nil {
+		return fmt.Errorf("could not calculate potentially accrued fees: %w", err)
+	}
+
+	// Invariant: Proposal transactions do not impact SoV state.
+	sovIterator, err := diff.GetActiveSubnetOnlyValidatorsIterator()
+	if err != nil {
+		return fmt.Errorf("could not iterate over active SoVs: %w", err)
+	}
+
+	var sovsToDeactivate []state.SubnetOnlyValidator
+	for sovIterator.Next() {
+		sov := sovIterator.Value()
+		// If the validator has exactly the right amount of fee for the next
+		// second we should not remove them here.
+		//
+		// GetActiveSubnetOnlyValidatorsIterator iterates in order of increasing
+		// EndAccumulatedFee, so we can break early.
+		if sov.EndAccumulatedFee >= potentialAccruedFees {
+			break
+		}
+
+		sovsToDeactivate = append(sovsToDeactivate, sov)
+	}
+
+	// The iterator must be released prior to attempting to write to the
+	// diff.
```

Comment:
nit: Could we also add the reason to the comment?

Reply:
I added the comment to the `GetActiveSubnetOnlyValidatorsIterator` definition on the `SubnetOnlyValidators` interface. It is not safe to modify this iterator because it internally uses an `interator.tree` - which has a race condition if the tree is modified during iteration.